File: app/src/test/java/ch/threema/app/utils/AppRestrictionUtilTest.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2019-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.utils;

import android.content.Context;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

@SuppressWarnings("ArraysAsListWithZeroOrOneArgument")
@RunWith(PowerMockRunner.class)
@PrepareForTest(AppRestrictionUtil.class)
public class AppRestrictionUtilTest {

    private void assertWebHosts(@Nullable String stringVal, @Nullable List<String> expected) {
        // Mock getStringRestriction method
        PowerMockito.stub(
            PowerMockito.method(AppRestrictionUtil.class, "getStringRestriction", String.class)
        ).toReturn(stringVal);

        // Mock context
        final Context context = Mockito.mock(Context.class);

        // Parse web hosts
        final List<String> result = AppRestrictionUtil.getWebHosts(context);
        if (expected == null) {
            Assert.assertNull(result);
        } else if (result == null) {
            Assert.fail("Result is null, but expected list of strings");
        } else {
            final Object[] expectedArray = expected.toArray();
            final Object[] resultArray = result.toArray();
            Assert.assertArrayEquals(expectedArray, resultArray);
        }
    }

    @Test
    public void getWebHostsNull() {
        this.assertWebHosts(null, null);
    }

    @Test
    public void getWebHostsEmpty() {
        this.assertWebHosts("", null);
    }

    @Test
    public void getWebHostsSingle() {
        this.assertWebHosts("saltyrtc.threema.ch", Arrays.asList("saltyrtc.threema.ch"));
    }

    @Test
    public void getWebHostsEmptyList() {
        this.assertWebHosts(",", null);
    }

    @Test
    public void getWebHostsMultiple() {
        this.assertWebHosts(
            "saltyrtc.threema.ch,test.threema.ch,*.example.com",
            Arrays.asList("saltyrtc.threema.ch", "test.threema.ch", "*.example.com")
        );
    }

    @Test
    public void getWebHostsFilterEmpty() {
        this.assertWebHosts(
            "saltyrtc.threema.ch,,foo.bar",
            Arrays.asList("saltyrtc.threema.ch", "foo.bar")
        );
    }

    @Test
    public void getWebHostsTrim() {
        this.assertWebHosts("  saltyrtc.threema.ch  ,, ", Arrays.asList("saltyrtc.threema.ch"));
    }

    private void assertWebHostAllowed(@Nullable List<String> whitelist, @NonNull String hostname, boolean allowed) {
        // Mock getWebHosts method
        PowerMockito.stub(
            PowerMockito.method(AppRestrictionUtil.class, "getWebHosts", Context.class)
        ).toReturn(whitelist);

        // Mock context
        final Context context = Mockito.mock(Context.class);

        // Validate hostname
        Assert.assertEquals(allowed, AppRestrictionUtil.isWebHostAllowed(context, hostname));
    }

    @Test
    public void isWebHostAllowedNo() {
        assertWebHostAllowed(Arrays.asList("example.com"), "threema.ch", false);
        assertWebHostAllowed(Arrays.asList("example.com"), "x.example.com", false);
        assertWebHostAllowed(Arrays.asList("*.example.com"), "x.example", false);
    }

    @Test
    public void isWebHostAllowedNoEmptyList() {
        assertWebHostAllowed(new ArrayList<>(), "example.com", false);
    }

    @Test
    public void isWebHostAllowedYesExact() {
        assertWebHostAllowed(Arrays.asList("example.com"), "example.com", true);
        assertWebHostAllowed(Arrays.asList("example.com", "x.example.com"), "x.example.com", true);
    }

    @Test
    public void isWebHostAllowedYesNullList() {
        assertWebHostAllowed(null, "example.com", true);
    }

    @Test
    public void isWebHostAllowedYesPrefixMatch() {
        assertWebHostAllowed(Arrays.asList("example.com", "*.example.com"), "x.example.com", true);
        assertWebHostAllowed(Arrays.asList("example.com", "*.example.com"), "xyz.example.com", true);
        assertWebHostAllowed(Arrays.asList("example.com", "*.example.com"), "x.y.example.com", true);
    }

}

File: app/src/test/java/ch/threema/app/utils/AppRestrictionUtilTest.kt
package ch.threema.app.utils

import ch.threema.app.restrictions.AppRestrictionUtil
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import io.mockk.unmockkStatic
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.fail

class AppRestrictionUtilTest {

    @BeforeTest
    fun setUp() {
        mockkStatic(AppRestrictionUtil::class)
    }

    @AfterTest
    fun tearDown() {
        unmockkStatic(AppRestrictionUtil::class)
    }

    @Test
    fun webHostsNull() {
        assertWebHosts(value = null, expected = null)
    }

    @Test
    fun webHostsEmpty() {
        assertWebHosts(value = "", expected = null)
    }

    @Test
    fun webHostsSingle() {
        assertWebHosts(value = "saltyrtc.threema.ch", expected = listOf("saltyrtc.threema.ch"))
    }

    @Test
    fun webHostsEmptyList() {
        assertWebHosts(value = ",", expected = null)
    }

    @Test
    fun webHostsMultiple() {
        assertWebHosts(
            value = "saltyrtc.threema.ch,test.threema.ch,*.example.com",
            expected = listOf("saltyrtc.threema.ch", "test.threema.ch", "*.example.com"),
        )
    }

    @Test
    fun webHostsFilterEmpty() {
        assertWebHosts(
            value = "saltyrtc.threema.ch,,foo.bar",
            expected = listOf("saltyrtc.threema.ch", "foo.bar"),
        )
    }

    @Test
    fun webHostsTrim() {
        assertWebHosts(value = "  saltyrtc.threema.ch  ,, ", expected = listOf("saltyrtc.threema.ch"))
    }

    @Test
    fun isWebHostAllowedNo() {
        assertWebHostAllowed(whitelist = listOf("example.com"), hostname = "threema.com", allowed = false)
        assertWebHostAllowed(whitelist = listOf("example.com"), hostname = "x.example.com", allowed = false)
        assertWebHostAllowed(whitelist = listOf("*.example.com"), hostname = "x.example", allowed = false)
    }

    @Test
    fun isWebHostAllowedNoEmptyList() {
        assertWebHostAllowed(whitelist = emptyList(), hostname = "example.com", allowed = false)
    }

    @Test
    fun isWebHostAllowedYesExact() {
        assertWebHostAllowed(whitelist = listOf("example.com"), hostname = "example.com", allowed = true)
        assertWebHostAllowed(whitelist = listOf("example.com", "x.example.com"), hostname = "x.example.com", allowed = true)
    }

    @Test
    fun isWebHostAllowedYesNullList() {
        assertWebHostAllowed(whitelist = null, hostname = "example.com", allowed = true)
    }

    @Test
    fun isWebHostAllowedYesPrefixMatch() {
        assertWebHostAllowed(whitelist = listOf("example.com", "*.example.com"), hostname = "x.example.com", allowed = true)
        assertWebHostAllowed(whitelist = listOf("example.com", "*.example.com"), hostname = "xyz.example.com", allowed = true)
        assertWebHostAllowed(whitelist = listOf("example.com", "*.example.com"), hostname = "x.y.example.com", allowed = true)
    }

    private fun assertWebHosts(value: String?, expected: List<String>?) {
        // Mock getStringRestriction method
        every { AppRestrictionUtil.getStringRestriction(any()) } returns value

        // Parse web hosts
        val result = AppRestrictionUtil.getWebHosts(
            mockk {
                every { getString(any()) } returns ""
            },
        )
        when {
            expected == null -> assertNull(result)
            result == null -> fail("Result is null, but expected list of strings")
            else -> assertContentEquals(expected, result)
        }
    }

    private fun assertWebHostAllowed(whitelist: List<String>?, hostname: String, allowed: Boolean) {
        // Mock getWebHosts method
        every { AppRestrictionUtil.getWebHosts(any()) } returns whitelist

        // Validate hostname
        assertEquals(allowed, AppRestrictionUtil.isWebHostAllowed(mockk(), hostname))
    }
}
